### PR TITLE
Ref updates for 7.2

### LINKED ISSUE #3241: Create new topics for ref features in C# 7.2
Bundling these features together because the explanation and the discussion will encompass all of them.  The related features are:

- [ref readonly parameters and returns](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/readonly-ref.md)
- readonly struct
- [stackalloc span](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md)
- ref extensions
- ref ternary (conditional operator)
- ref struct

Note: this issue will be updated with the other specs as I find updated references to them. 

Other background on the feature is [here](https://github.com/VSadov/csharplang/blob/RefSpanDesignChoices/proposals/RefReadonlySpanDesignChoices.md)

@VSadov is one lead for these features.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #3241 

Explain the concepts behind `in` parameters, `ref readonly` returns, `readonly struct` types and `ref struct` types.

### CODE REVIEW

**Comment by BillWagner** [comment]
I checked the build and the links, this is ready to review.

/cc @VSadov

File: docs/csharp/ref-locals-and-returns.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+You may also want to return a value type by reference, but disallow the
+caller from modifying that value. Use the `ref readonly` modifier to
+express that design intent. It notifies readers
+that you are returning a reference to existing data, but not allowing
+modification. 
+
+The compiler enforces that the caller cannot modify the reference. Attempts
+to assign to the value directly generate a compile-time error. Calling member
+methods creates a defensive copy.
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: define defensive copy

File: docs/csharp/ref-locals-and-returns.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+There are several ways in which the compiler ensures that the readonly
+nature of an `in` argument is enforced.  First of all, the called method
+can't directly assign to an `in` parameter. It can't directly assign
+to any field of an `in` parameter. In addition, you cannot pass
+an `in` parameter to any method demanding the `ref` or `out` modifier.
+The compiler enforces that the `in` argument is a readonly variable. You
+can call any instance method that uses pass-by-value semantics. In
+those instances, a copy of the `in` parameter is created. Because the compiler
+can create a temporary for any `in` parameter, you can also specify default
+values for any `in` parameter. The follow code uses that to specify the origin
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: define what origin is here, just to make it clear -- i.e, origin (point 0,0).

File: docs/csharp/ref-locals-and-returns.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+methods creates a defensive copy. 
+
+Creating a copy of a ref readonly return is easy: Just assign it to a variable
+not declared with the `ref readonly` modifier. The compiler generates code
+to copy the object as part of the assignment. 
+
+It's likely that the library using `Point3D` would often use the origin
+throughout the code. Every instance creates a new object on the stack. It may
+be advantageous to create a constant and return it by reference. But, if you
+are returning a reference to internal storage, you may want to enforce that
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: are returning --> return

File: docs/csharp/ref-locals-and-returns.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+caller from modifying that value. Use the `ref readonly` modifier to
+express that design intent. It notifies readers
+that you are returning a reference to existing data, but not allowing
+modification. 
+
+The compiler enforces that the caller cannot modify the reference. Attempts
+to assign to the value directly generate a compile-time error. Calling member
+methods creates a defensive copy. 
+
+Creating a copy of a ref readonly return is easy: Just assign it to a variable
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: I'd move this to follow the example. It's best to have the example -- which is a good one -- follow immediately to provide a use case for ref readonly and make it less abstract.

File: docs/csharp/ref-locals-and-returns.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+throughout the code. Every instance creates a new object on the stack. It may
+be advantageous to create a constant and return it by reference. But, if you
+are returning a reference to internal storage, you may want to enforce that
+the caller cannot modify the referenced storage. The following code
+defines a read-only property that returns a `readonly ref` to a `Point3D`
+that specifies the origin.
+
+[!code-csharp[OriginReference](../../samples/csharp/reference-semantics/Point3D.cs#OriginReference "Creating a readonly Origin reference")]
+
+When you assign a variable to a `ref return`, you can specify either a `ref readonly`
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: is `ref return` `ref readonly return`?

File: docs/csharp/ref-locals-and-returns.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+variable, or a by-value copy of the readonly reference:
+
+[!code-csharp[AssignRefReadonly](../../samples/csharp/reference-semantics/Program.cs#AssignRefReadonly "Assigning a ref readonly")]
+
+The first assignment in the preceding code makes a copy of the `Origin` constant, and assigns
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: no comma after constant

File: docs/csharp/ref-locals-and-returns.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+that copy. The second assigns a reference. Notice that the `readonly` modifier
+must be part of the declaration of the variable. The reference to which it refers
+cannot be modified. Attempts to do so results in a compile-time error.
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: results --> result

File: docs/csharp/ref-locals-and-returns.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+## `readonly struct` type
+
+Applying `ref readonly` to high-traffic uses of a struct may be sufficient.
+Other times, you may want to create an immutable struct. Then, you can
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: no comma after then

File: docs/csharp/ref-locals-and-returns.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+always pass by readonly reference. That practice 
+removes the defensive copies
+that take place when you access methods of a struct used as an `in` parameter.
+
+You can do that by creating a `readonly struct` type. You can add the `readonly`
+modifier to a struct declaration. The language enforces that all members of
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: compiler enforces?

File: docs/csharp/ref-locals-and-returns.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+the struct are `readonly`: the `struct` must be immutable.
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: semicolon instead of colon.

File: docs/csharp/ref-locals-and-returns.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+use the `in` modifier at every location where a `readonly struct` is an
+argument. In addition, you can return a `readonly struct` as a `ref return` when
+you are returning an object whose lifetime extends beyond the scope of the method
+returning the object.
+
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
+of a copy of the receiver, passed by value to the member method. This optimization
+saves more copying when you use a `readonly struct`. The `Point3D` would be a great
+candidate for this change. The following code shows an update `ReadonlyPoint3D` structure:
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: update --> updated

File: docs/csharp/ref-locals-and-returns.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+the struct are `readonly`: the `struct` must be immutable.
+
+There are other optimizations for a `readonly struct`. You can
+use the `in` modifier at every location where a `readonly struct` is an
+argument. In addition, you can return a `readonly struct` as a `ref return` when
+you are returning an object whose lifetime extends beyond the scope of the method
+returning the object.
+
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: Hard to parse. Perhaps "The `this` reference, instead of a copy of the receiver, is always an `in` parameter passed by value to the member method.

File: docs/csharp/ref-locals-and-returns.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+of a copy of the receiver, passed by value to the member method. This optimization
+saves more copying when you use a `readonly struct`. The `Point3D` would be a great
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: nit: would be --> is

File: docs/csharp/ref-locals-and-returns.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+[!code-csharp[ReadonlyOnlyPoint3D](../../samples/csharp/reference-semantics/Point3D.cs#ReadonlyOnlyPoint3D "Defining an immutable structure")]
+
+## `ref struct` type
+
+Another related language feature is the ability to declare a value type that
+must be stack allocated. In other words, these types can never be created on the
+heap as a member of another class. The primary motivation for this feature
+was `Span<T>` and related structures, but you can create `ref struct` types
+when you have similar requirements. In this article, you see examples
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: The requirements aren't clear. Can you mention a use case here?

File: docs/csharp/ref-locals-and-returns.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+`Span<T>` would be subject to out of range errors or other type safety violations
+were it not constrained to a single stack frame. In addition, putting a
+managed pointer on the GC heap typically crashes at JIT time.
+
+The `ref struct` declaration declares that a struct of this type must be
+on the stack. The language rules ensure the safe use of these types. Other
+types declared as `ref struct` include `ReadOnlySpan<T>`, `Memory<T>`, and
+`ReadOnlyMemory<T>`. 
+
+The goal of keeping a `ref struct` type as a stack allocated variable
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: stack allocated --> stack-allocated

File: docs/csharp/ref-locals-and-returns.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+These restrictions ensure that you do not accidentally use a `ref struct`
+in a manner that could promote it to the managed heap.
+
+## Conclusions
+
+These enhancements to the C# language are designed for performance critical
+algorithms where memory allocations can be critical to achieving the
+necessary performance. You may find that you don't often use these features in
+the code you write. However, these enhancements have been adopted in many locations
+in the .NET framework. As more and more APIs make use of these features, you'll
```
Comment (rpetrusha) [inline docs/csharp/ref-locals-and-returns.md]: framework --> Framework

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This looks good. Left a number of comments, @BillWagner.

File: docs/csharp/ref-locals-and-returns.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+This addition provides a full vocabulary to express your design intent:
+
+- `out`: This method sets the value of the argument used as this parameter.
+- `ref`: This method may set the value of the argument used as this parameter.
+- `in`: This method does not modify the value of the argument used as this parameter.
+
+When you add the `in` modify to pass an argument by reference, you declare
+your design intent is to pass arguments by reference to
+avoid unnecessary copying. You do not intend to modify the object used
+as that argument. The following code shows an example of a method
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/ref-locals-and-returns.md]: Good point. I added more context in the paragraph above.

File: docs/csharp/ref-locals-and-returns.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/ref-locals-and-returns.md]: I updated the paragraph to define it.

File: docs/csharp/ref-locals-and-returns.md (hunk 12 above)
Comment (BillWagner) [reply docs/csharp/ref-locals-and-returns.md]: Also, it should be "passed by reference" not "passed by value".

File: docs/csharp/ref-locals-and-returns.md (hunk 14 above)
Comment (BillWagner) [reply docs/csharp/ref-locals-and-returns.md]: I re-wrote that section. Likely, using `Span<T>` and related are the most common use case. I tried to make that point, and point to other uses.

**Comment by BillWagner** [comment]
@rpetrusha I updated everything based on your comments. Let me know if this is ready.

I also changed the file name in a separate commit so it should be easier to review.

File: docs/csharp/reference-semantics-with-value-types.md [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+you can configure the language version. Choose either "7.2", or "latest".  Or you can
+edit the *csproj* file and add the following node:
+
+```XML
+  <PropertyGroup>
+    <LangVersion>7.2</LangVersion>
+  </PropertyGroup>
+```
+
+You can use either "7.2" or "latest" for the value.
```
Comment (jcouv) [inline docs/csharp/reference-semantics-with-value-types.md]: Given that there are multiple ways of updating a project (UI settings, editing csproj, code fix), consider factoring this section to a dedicated page, and referenced from here.
Then you can just have one line: "In order to use those features, you have to configure your project to use C# 7.2 or later in your project, by setting the language version to "7.2" or "latest" per _these instructions_".

Comment (BillWagner) [reply docs/csharp/reference-semantics-with-value-types.md]: I'd agree. 

I want to take a bit more time to do that, so I made an issue that can get addressed in this or the next sprint: #3685

**Review by rpetrusha — APPROVED** [review]
LGTM, @BillWagner. This is good to merge when you're ready, after the build completes.
